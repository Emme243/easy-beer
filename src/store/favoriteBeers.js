export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
