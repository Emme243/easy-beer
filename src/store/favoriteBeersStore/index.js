export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
