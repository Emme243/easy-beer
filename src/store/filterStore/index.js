import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
