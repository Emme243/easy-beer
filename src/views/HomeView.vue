<template>
  <v-app>
    <v-container class="text-center mt-5 body-1">
      <TheHeader />
      <TheNavbar />
      <v-main>
        <BeerList />
      </v-main>
      <Pagination />
    </v-container>
  </v-app>
</template>

<script>
import BeerList from '@/components/BeerList';
import Pagination from '@/components/BeerFilters/Pagination';
import TheHeader from '@/components/UI/TheHeader';
import TheNavbar from '@/components/UI/TheNavbar';
import { mapActions } from 'vuex';

export default {
  name: 'HomeView',
  components: {
    BeerList,
    Pagination,
    TheHeader,
    TheNavbar,
  },
  methods: mapActions({
    setAllFiltersInStore: 'filter/setAllFilters',
    setFavoriteBeersInStore: 'favoriteBeers/setFavoriteBeers',
  }),
  mounted() {
    // NOTE: En este punto ya se ha cargado el componente BeerLayout.vue
    //       y se puede acceder a los datos de filtrado en la query de la url,
    //       por lo que podemos setear esos valores en el store.
    //       De este modo, cada vez que el usuario refresque la página, no perderá los filtros.
    this.setAllFiltersInStore(this.$route.query);

    // NOTE: También podemos traer todas las cervezas marcadas como favoritas
    //       por el usuario desde el localStorage y setearlas en el store.
    const favoriteBeers = localStorage.getItem('favoriteBeers');
    if (favoriteBeers) this.setFavoriteBeersInStore(JSON.parse(favoriteBeers));
  },
};
</script>
