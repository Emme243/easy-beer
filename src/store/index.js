import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
