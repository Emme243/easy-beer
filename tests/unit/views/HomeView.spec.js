import store from '@/store';
import HomeView from '@/views/HomeView';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import { beersMock } from '../mocks/beersMock';

describe('Pruebas en el componente HomeView.vue', function () {
  let vuetify;
  const localVue = createLocalVue();
  const initialQueryParams = {
    page: 5,
    searchQuery: 'Cerveza linda',
    minAbvValue: 5,
    initialBrewedMonth: '01-2020',
  };

  const mountFunction = () => {
    return shallowMount(HomeView, {
      localVue,
      vuetify,
      store,
      mocks: {
        $route: { query: initialQueryParams },
      },
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
    store.dispatch = jest.fn();
  });

  it('Debe hacer match con el snapshot', function () {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Debe haber llamado el dispatch setAllFilters con los query params del url al haberse montado', function () {
    mountFunction();
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith('filter/setAllFilters', initialQueryParams);
  });

  it('No debe llamar al dispatch setFavoriteBeers si no hay cervezas favoritas en el localStorage', function () {
    mountFunction();
    expect(store.dispatch).not.toHaveBeenCalledWith('favoriteBeers/setFavoriteBeers');
    expect(store.dispatch).toHaveBeenCalledTimes(1); // Se llama al dispatch setAllFilters
  });

  it('Debe llamar al dispatch setFavoriteBeers si hay cervezas favoritas en el localStorage', function () {
    localStorage.setItem('favoriteBeers', JSON.stringify(beersMock));
    mountFunction();
    expect(store.dispatch).toHaveBeenCalledWith('favoriteBeers/setFavoriteBeers', beersMock);
    expect(store.dispatch).toHaveBeenCalledTimes(2); // Se llama al dispatch setAllFilters y al dispatch setFavoriteBeers
  });
});
