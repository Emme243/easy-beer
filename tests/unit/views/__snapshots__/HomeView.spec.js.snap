// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente HomeView.vue Debe hacer match con el snapshot 1`] = `
<v-app-stub id="app">
  <v-container-stub tag="div" class="text-center mt-5 body-1">
    <theheader-stub></theheader-stub>
    <thenavbar-stub></thenavbar-stub>
    <v-main-stub tag="main">
      <beerlist-stub></beerlist-stub>
    </v-main-stub>
    <pagination-stub></pagination-stub>
  </v-container-stub>
</v-app-stub>
`;
